File: mac-app/QuartzTeachers/Scripting/NewCourseCreator.swift
import Foundation
import Observation

/// Creates a new course by driving the real setup wizard.
///
/// The app writes the teacher's chosen settings to
/// `courses/<CODE>/course_config.json` first, then runs `./setup.sh`. The
/// wizard treats that file as its saved answers and offers each one as a
/// default, so this type simply watches the output and "presses Return" at
/// every prompt (typing the course code at the one prompt that needs it).
/// All folder scaffolding, backups, and Quartz patching happen inside the
/// real script — the app adds no logic of its own.
@Observable
class NewCourseCreator {

    // MARK: - Stored properties

    let runner: ScriptRunner = ScriptRunner()

    /// True from start until setup.sh finishes.
    var isCreating: Bool = false

    /// A problem preparing the run (before the script started), if any.
    var preparationProblem: String?

    /// The course code being created, uppercased.
    private var courseCode: String = ""

    /// How much of the transcript the pump has already responded to.
    private var respondedLength: Int = 0

    /// Safety valve so a confused run cannot send input forever.
    private var responsesSent: Int = 0

    // MARK: - Functions

    /// Writes the config, then starts `setup.sh` and the answer pump.
    func createCourse(configuration: [String: Any], workspaceURL: URL) {
        preparationProblem = nil

        guard let storedCode = configuration["course_code"] as? String else {
            preparationProblem = "The course needs a code."
            return
        }
        courseCode = storedCode.uppercased()

        let courseDirectoryURL: URL = workspaceURL
            .appendingPathComponent("courses")
            .appendingPathComponent(courseCode)
        let configFileURL: URL = courseDirectoryURL.appendingPathComponent("course_config.json")

        do {
            try FileManager.default.createDirectory(at: courseDirectoryURL, withIntermediateDirectories: true)
            let options: JSONSerialization.WritingOptions = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
            var data: Data = try JSONSerialization.data(withJSONObject: configuration, options: options)
            data.append(contentsOf: [0x0A])
            try data.write(to: configFileURL, options: [.atomic])
        } catch {
            preparationProblem = "Could not write the course configuration: \(error.localizedDescription)"
            return
        }

        respondedLength = 0
        responsesSent = 0
        isCreating = true
        runner.run(scriptNamed: "setup.sh", arguments: [], workingDirectory: workspaceURL)

        Task {
            await pumpAnswers()
        }
    }

    /// Watches the wizard's output; whenever a prompt is waiting, answers
    /// it — the course code at the code prompt, Return everywhere else.
    private func pumpAnswers() async {
        while runner.isRunning {
            try? await Task.sleep(for: .milliseconds(400))

            if responsesSent > 300 {
                // Something is looping; stop feeding it.
                runner.terminate()
                break
            }

            let displayText: String = runner.transcript.displayText
            if displayText.count == respondedLength {
                // No new output since our last response; the script is
                // either busy working or waiting at a prompt we already
                // answered. Only answer when NEW output ends in a prompt.
                continue
            }

            let lastLine: String = currentPromptLine(in: displayText)
            if !looksLikePrompt(lastLine) {
                continue
            }

            respondedLength = displayText.count
            responsesSent += 1

            if lastLine.contains("Enter the course code") {
                runner.send(line: courseCode)
            } else {
                runner.send(line: "")
            }
        }
        isCreating = false
    }

    /// The last non-empty line of output — the prompt currently waiting.
    private func currentPromptLine(in text: String) -> String {
        let lines: [Substring] = text.split(separator: "\n", omittingEmptySubsequences: false)
        var index: Int = lines.count - 1
        while index >= 0 {
            let line: String = String(lines[index]).trimmingCharacters(in: .whitespaces)
            if !line.isEmpty {
                return line
            }
            index -= 1
        }
        return ""
    }

    /// Heuristics for "the wizard is waiting for input on this line" —
    /// matched to the actual prompt shapes in setup_course.py.
    private func looksLikePrompt(_ line: String) -> Bool {
        if line.hasSuffix(":") {
            return true
        }
        if line.hasSuffix(": ") {
            return true
        }
        if line == ">" {
            return true
        }
        if line.hasPrefix("Use ← / →") {
            // The colour scheme picker: Return accepts the saved default.
            return true
        }
        if line.hasSuffix("?") {
            return true
        }
        return false
    }
}

File: windows-app/Plantoir.Core/Scripting/NewCourseCreator.cs
using System;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;
using Plantoir.Core.Models;

namespace Plantoir.Core.Scripting;

/// <summary>
/// "The wizard with every answer pre-filled": the app writes the teacher's
/// choices to course_config.json first, runs the REAL setup launcher, and
/// presses Return at each prompt — the wizard re-reads the file as its
/// saved defaults, so a plain Return accepts every one. The one exception,
/// the course-code prompt, is answered explicitly.
/// </summary>
public sealed class NewCourseCreator
{
    public ScriptRunner Runner { get; }
    public bool IsCreating { get; private set; }
    public string? PreparationProblem { get; private set; }
    public string? InstalledExampleCode { get; private set; }

    private string _courseCode = "";
    private long _respondedVersion = -1;
    private int _responsesSent;

    public NewCourseCreator(ScriptRunner runner) => Runner = runner;

    public async Task CreateCourse(JObject configuration, string workspacePath)
    {
        PreparationProblem = null;
        if (configuration["course_code"]?.ToString() is not { Length: > 0 } storedCode)
        {
            PreparationProblem = "The course needs a code.";
            return;
        }
        _courseCode = storedCode.ToUpperInvariant();

        try
        {
            string courseDir = Path.Combine(Workspace.CoursesDirectory(workspacePath), _courseCode);
            Directory.CreateDirectory(courseDir);
            CourseConfiguration.FromDictionary(configuration)
                .Write(Path.Combine(courseDir, "course_config.json"));
        }
        catch (Exception error)
        {
            PreparationProblem = $"Could not write the course configuration: {error.Message}";
            return;
        }

        _respondedVersion = -1;
        _responsesSent = 0;
        IsCreating = true;
        Runner.Milestones = TaskMilestones.CourseCreation;
        Runner.Run("setup.ps1", Array.Empty<string>(), workspacePath);
        await PumpAnswers();
        IsCreating = false;
    }

    public async Task InstallExampleCourse(string workspacePath)
    {
        PreparationProblem = null;
        IsCreating = true;
        Runner.Milestones = TaskMilestones.ExampleCourse;
        Runner.Run("setup.ps1", new[] { "--install-example" }, workspacePath);
        await Runner.WaitUntilFinished();
        InstalledExampleCode = OutputParsers.ExampleCourseCode(Runner.Transcript.DisplayText);
        IsCreating = false;
    }

    /// <summary>
    /// Every 400 ms: if NEW output has arrived and its last non-empty line is
    /// prompt-shaped, answer — the course code where asked, a bare Return
    /// everywhere else. The new-output guard is the whole anti-double-answer
    /// mechanism; 300 responses is the runaway valve.
    /// </summary>
    /// <summary>
    /// Wait for output to go quiet before answering, so we only ever act on a
    /// COMPLETE prompt. Without this, a multi-line render whose intermediate
    /// lines happen to end in ':' — the colour-scheme picker prints
    /// "Light Mode:" and "Dark Mode:" before its "Use ← / →" footer — would be
    /// answered mid-render, firing an extra Return into the raw key reader and
    /// desyncing the whole sequence into a deadlock.
    /// </summary>
    private const int SettleMilliseconds = 600;

    private async Task PumpAnswers()
    {
        while (Runner.IsRunning)
        {
            await Task.Delay(200);
            if (_responsesSent > 300) { Runner.Terminate(); break; }
            long version = Runner.Transcript.Version;
            if (version == _respondedVersion) continue;
            // Output is still streaming — let the current prompt finish printing.
            if ((DateTime.UtcNow - Runner.LastOutputAt).TotalMilliseconds < SettleMilliseconds) continue;
            // This version is now settled; mark it handled either way so a
            // non-prompt (a status line between prompts) doesn't busy-loop.
            _respondedVersion = version;
            string lastLine = CurrentPromptLine();
            if (!LooksLikePrompt(lastLine)) continue;
            _responsesSent++;
            Runner.SendLine(lastLine.Contains("Enter the course code") ? _courseCode : "");
        }
    }

    private string CurrentPromptLine()
    {
        string current = Runner.Transcript.CurrentLine.Trim();
        if (current.Length > 0) return current;
        for (int i = Runner.Transcript.Lines.Count - 1; i >= 0; i--)
        {
            string line = Runner.Transcript.Lines[i].Trim();
            if (line.Length > 0) return line;
        }
        return "";
    }

    /// <summary>
    /// Looser than the question heuristic on purpose — matched to the prompt
    /// shapes setup_course.py really prints, including the arrow-key colour
    /// picker (where Return accepts the current scheme).
    ///
    /// The picker footer is matched by CONTAINS, not StartsWith: the Windows
    /// pseudo console sometimes renders it onto the tail of the last swatch
    /// line ("… textHighlight Use ← / → … Enter to select …"), so anchoring at
    /// the start would miss it and the run would deadlock waiting for input
    /// that never comes.
    /// </summary>
    internal static bool LooksLikePrompt(string line) =>
        line.EndsWith(':') || line.EndsWith(": ", StringComparison.Ordinal)
        || line == ">"
        || line.Contains("Enter to select", StringComparison.Ordinal)
        || line.EndsWith('?');
}
